runtime/manifest-parser.peg

```diff
@@ -38,36 +41,186 @@ ManifestItem
   / Schema
 
 Import
-  = "import" whiteSpace path:id eolWhiteSpace
+  = 'import' whiteSpace path:id eolWhiteSpace
   {
     return {
-      kind: "import",
+      kind: 'import',
       path,
     }
   }
 
 Particle
-  = "particle" whiteSpace name:upperIdent implFile:(whiteSpace "in" whiteSpace id)? eolWhiteSpace body:IndentedText?
+  = 'particle' whiteSpace name:upperIdent implFile:(whiteSpace 'in' whiteSpace id)? eolWhiteSpace items:(Indent (SameIndent ParticleItem)*)? eolWhiteSpace?
   {
+    let args = null;
+    let affordance = [];
+    let slots = [];
+    let description = null;
+    items = items ? extractIndented(items) : [];
+    items.forEach(item => {
+      if (item.name == name) {
+        args = item.args;
+      } else if (item.kind == 'particle-slot') {
+        slots.push(item);
+      } else if (item.kind == 'description') {
+        description = {};
+        item.description.forEach(d => { description[d.name] = d.pattern; });
+      } else {
+        affordance.push(item.affordance)
+      }
+    });
+
     return {
-      kind: "particle",
+      kind: 'particle',
       name,
       implFile: optional(implFile, implFile => implFile[3], null),
-      body,
+      args,
+      affordance,
+      slots,
+      description
+    };
+  }
+
+ParticleItem
+  = ParticleInterface
+  / Affordance
+  / ParticleSlot
+  / ParticleDescription
+
+ParticleInterface
+  = name:upperIdent '(' args:ParticleArgumentList ')' eolWhiteSpace
+  {
+    return { name, args }
+  }
+
+ParticleArgumentList
+  = head:ParticleArgument tail:(',' whiteSpace ParticleArgument)*
+  {
+    return [head].concat(tail.map(a => a[2]));
+  }
+
+ParticleArgument
+  = direction:ParticleArgumentDirection whiteSpace tn:ParticleArgumentType whiteSpace name:(lowerIdent)
+  {
+    return {direction, type:tn, name};
+  }
+
+ParticleArgumentDirection
+  = 'in' / 'out' / 'inout'
+  {
+     // Do we still need ephemeral?
+     return {direction: dir};
+  }
+
+ParticleArgumentType
+  = (VariableType / PrimitiveType / ListType)
+ListType
+  = '[' pt:ParticleArgumentType ']' { return typeLiteral.viewOf(pt); }
+VariableType
+  = '~' name: VariableTypeName { return typeLiteral.namedTypeVariable(name); }
+VariableTypeName
+  = [a-z][a-zA-Z0-9]* { return text() }
+PrimitiveType
+  = [A-Z][a-zA-Z0-9]* { return text() }
+
+Affordance
+  = 'affordance' whiteSpace affordance:('dom-touch' / 'dom' / 'vr' / 'voice') eolWhiteSpace
+  { return {affordance}; }
+
+ParticleSlot
+  = isRequired:('must' whiteSpace)? 'consume' whiteSpace name:(lowerIdent) eolWhiteSpace
+    items:(Indent (SameIndent ParticleSlotItem)*)?
+  {
+    let formFactor = null;
+    let providedSlots = [];
+    items = items ? extractIndented(items) : [];
+    items.forEach(item => {
+      if (item.kind == 'provided-slot') {
+        providedSlots.push(item);
+      } else if (item.kind == 'form-factor') {
+        if (formFactor)
+          assert ('duplicate form factor for a slot');
```

Comment:
Can you try to call `error('duplicate form...` here? That should cause the parser to throw it's regular error object with position information.